### PR TITLE
JDK-8254189: Improve comments for StackOverFlow and fix in_xxx() functions

### PR DESCRIPTION
**Pull request by tstuefe**
Hi,

may I please have reviews for this small cleanup / fix?

While reviewing JDK-8253717 it was found that comments would help with understanding the StackOverFlow class. Especially the fact that the various _base_ are actually pointing outside their respective zone, since the stack grows downward and a zone (and the stack itself) range is [end, base). If you don't look at this code daily it can be surprising.

This also fixes some small off-by-one errors in various "in_stack_xxx_zone()" methods which test whether a given address is inside a zone and gave wrong results for address=base since base points outside its zone. This had the effect that an address could be in multiple zones.

Finally it adds a small gtest which tests the StackOverFlow methods.

/issue: JDK-8254189

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Testing

|     | Linux x32 | Linux x64 | Windows x64 | macOS x64 |
| --- | ----- | ----- | ----- | ----- |
| Build | ✔️ (1/1 passed) | ✔️ (5/5 passed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) |
| Test (tier1) |    |  ✔️ (9/9 passed) | ❌ (1/9 failed) | ✔️ (9/9 passed) |

**Failed test task**
- [Windows x64 (hs/tier1 compiler)](https://github.com/tstuefe/jdk/runs/1322483258)

### Issue
 * [JDK-8254189](https://bugs.openjdk.java.net/browse/JDK-8254189): Improve comments for StackOverFlow and fix in_xxx() functions

### Reviewers
 * [David Holmes](https://openjdk.java.net/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Gerard Ziemski](https://openjdk.java.net/census#gziemski) (@gerard-ziemski - Committer) ⚠️ Review applies to cc54ef8d732d86236a483738f082a89836ebe92a

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/795/head:pull/795`
`$ git checkout pull/795`

### CODE REVIEW

File: src/hotspot/share/runtime/stackOverflow.cpp
Comment (dholmes-ora) [inline src/hotspot/share/runtime/stackOverflow.cpp:44]: s/regardless the/regardless of the/

But this is already covered by the following comment:

  // We need to adapt the configured number of stack protection pages given
  // in 4K pages to the actual os page size. We must do this before setting

**Review by dholmes-ora — APPROVED** [review]
Hi Thomas,

Seems okay.

One nit below.

Thanks,
David

File: src/hotspot/share/runtime/stackOverflow.cpp
Comment (tstuefe) [reply src/hotspot/share/runtime/stackOverflow.cpp:44]: Okay, I'll remove the comment.

**Review by dholmes-ora — APPROVED** [review]

**Comment by tstuefe** [comment]
Ping... May I have a second review please? Its quite trivial.
Patch was tested in our internal systems and no problems surfaced.

**Comment by gerard-ziemski** [comment]
Very nice.

Personally I'd like to see additional:

  ASSERT_TRUE(so.in_stack_red_zone(so.stack_end()));
  ASSERT_TRUE(so.in_stack_yellow_zone(so.stack_red_zone_base()));
  ASSERT_TRUE(so.in_stack_reserved_zone(so.stack_yellow_zone_base()));

in the test, to drive home the point of the inclusion of the ranges, but that's a tiny nick pick.

**Review by gerard-ziemski — APPROVED** [review]

**Comment by tstuefe** [comment]
> Very nice.

Thank you, Gerard.

> 
> Personally I'd like to see additional:
> 
> ASSERT_TRUE(so.in_stack_red_zone(so.stack_end()));
> ASSERT_TRUE(so.in_stack_yellow_zone(so.stack_red_zone_base()));
> ASSERT_TRUE(so.in_stack_reserved_zone(so.stack_yellow_zone_base()));
> 
> in the test, to drive home the point of the inclusion of the ranges, but that's a tiny nick pick.

I can do this. Will have to reshape those a bit since there is no "in_yellow_zone" (I considered adding one for tests sake but then did not).

...Thomas

**Comment by tstuefe** [comment]
... and StackOverFlow::stack_end() is private, so I cannot use it in ASSERT either...

**Review by dholmes-ora — APPROVED** [review]

**Comment by coleenp** [comment]
Sorry for not getting to this review on time.  Thank you for doing this and writing the gtest.  This is really nice.

**Pull request closed without merging**
